nxos_static_route: route removal fails when track config is present

chrisvanheuveln opened the pull request:
##### SUMMARY
* `reconcile_candidate()`
  * old code searched the ip route configs for a given prefix+nexthop and then tried to remove the route based on prefix+nexthop only; this would fail when a static route was configured with `track` values.
  * new code still looks for prefix+nexthop but uses the route config it finds on the device to remove it; e.g.
    * search for: `ip route 192.168.20.64/24 192.0.2.3`
    * find:       `ip route 192.168.20.64/24 192.0.2.3 track 1 10`
    * remove:  `no ip route 192.168.20.64/24 192.0.2.3 track 1 10`

* logic cleanups:
  * old code did a `show run` for every prefix. This can be a lot of data when there are large configs.
  * new code uses filters to only return the static route configs.
  * The filters now allow a common code path so no need for default vs vrf code paths

* `sanity` test: 100% Pass rate on N9K,N7K,N6K,N3K

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
`nxos_static_route`

chrisvanheuveln committed:
nxos_static_route: route removal fails when track config is present

* `reconcile_candidate()`
 * old code searched the ip route configs for a given prefix+nexthop and then tried to remove the route based on prefix+nexthop only; this would fail when a static route was configured with `track` values.
 * new code still looks for prefix+nexthop but uses the route config it finds on the device to remove it; e.g.
  * search for: `ip route 192.168.20.64/24 192.0.2.3`
  * find:       `ip route 192.168.20.64/24 192.0.2.3 track 1 10`
  * remove:  `no ip route 192.168.20.64/24 192.0.2.3 track 1 10`

* logic cleanups:
 * old code did a `show run` for every prefix. This can be a lot of data when there are large configs.
 * new code uses filters to only return the static route configs.
 * The filters now allow a common code path so no need for default vs vrf code paths

* `sanity` test: 100% Pass rate on N9K,N7K,N6K,N3K

chrisvanheuveln committed:
nxos_static_route: route removal fails when track config is present

* `reconcile_candidate()`
 * old code searched the ip route configs for a given prefix+nexthop and then tried to remove the route based on prefix+nexthop only; this would fail when a static route was configured with `track` values.
 * new code still looks for prefix+nexthop but uses the route config it finds on the device to remove it; e.g.
  * search for: `ip route 192.168.20.64/24 192.0.2.3`
  * find:       `ip route 192.168.20.64/24 192.0.2.3 track 1 10`
  * remove:  `no ip route 192.168.20.64/24 192.0.2.3 track 1 10`

* logic cleanups:
 * old code did a `show run` for every prefix. This can be a lot of data when there are large configs.
 * new code uses filters to only return the static route configs.
 * The filters now allow a common code path so no need for default vs vrf code paths

* `sanity` test: 100% Pass rate on N9K,N7K,N6K,N3K

bcoca committed:
Fine tune sanity (#53544)

* modify regex to use implicit charsets this should solve issues in py3 and unicode names
* fix issue with subgroups in yaml inventory
* clarify deprecation message
* separated per name warning from deprecation
* move noise to verbosity, simplify warnings
* fix docs to reflect actual 'good' practice
* change toggle to choice list to give users more options

tsavery-DD committed:
nso documentation fixes (#53478)

chrisvanheuveln committed:
SA fixes for automatic field numbering

chrisvanheuveln committed:
SA fixes for automatic field numbering

richlv committed:
add warnings for the stding usage examples (#51859)

* Add a warning about vault leaving secrets in the shell history.
* Add a warning about accidental newlines in vault encrypted strings.

+label: docsite_pr

warusadura committed:
redfish_utils: add support to automate multiple systems. (#50854)

* Add support to automate multiple systems

Currently `redfish_utils` module can support only one system.
These changes introduce the ability to support multiple systems.

* Update - add support to automate multiple systems

* Introduce `aggregate()` to avoid code duplication in `get_multi___` like
methods.

* Remove "set" related methods

* include systems URI in each aggregated entry

* rebased

* Fixes KeyError and merge errors

jborean93 committed:
Add arg and doc validation for PowerShell modules (#53615)

* Add arg and doc validation for PowerShell modules

* Verify if pwsh exists before running it

smnmtzgr committed:
new module aci_access_sub_port_block_to_access_port (#51940)

* new module aci_access_sub_port_block_to_access_port

* new line at end of file

sugitk committed:
added `pmrun` and `enable` to `become-method` (#51009)

* added `pmrun` and `enable` to `become-method`

* references become plugin list for 'become_method' options

anis016 committed:
Updated playbook works sample. (#51673)

<!--- Your description here -->

Removed `|` from the playbook works sample.

+label: docsite_pr

mattclay committed:
Add support for ansible-test --coverage-check. (#53573)

mattclay committed:
Fix ansible-test coverage stub memory usage. (#53652)

Split coverage stub files into multiple parts based on the number
of lines in files included in each stub. This will keep memory usage
down to avoid hanging on Shippable.

UnknownGnome committed:
Update vmware_vmkernel.py (#53084)

bcoca committed:
remove obsolete set_play on play_context (#47934)

* remove redundant assignments from play

 - fix unit tests

jborean93 committed:
win_psexec - deprecate the extra_opts module options (#53666)

jokajak committed:
Include Windows as a platform in playbooks_conditionals (#52155)

Update the playbooks_conditionals page to include Windows as a valid platform for os_family

Co-Authored-By: jokajak <jokajak@gmail.com>

sdodsley committed:
Fix typos in purefa_pgsnap module (#53643)

mattclay committed:
Fix integration tests with custom inventory.

All inventory hosts to which a connection is actually established
should declare ansible_python_interpreter to avoid interpreter
auto-discovery, as that may select the wrong interpreter.

jborean93 committed:
Use unicode string for auto interpreter warnings (#53671)

* Use unicode string for auto interpreter warnings

* Added some unit tests for interpreter selection

* Fix python 3 syntax issues

mattclay committed:
Fix integration test inventory.

migumun committed:
Fortinet's FortiOS web proxy profile (#52843)

* Fortinet's FortiOS web proxy profile

* updated with changes from comments

migumun committed:
Fortinet's FortiOS web proxy global (#52842)

* Fortinet's FortiOS web proxy global

* Updated with changes from comments

migumun committed:
Fortinet's FortiOS web proxy explicit (#52841)

* Fortinet's FortiOS web proxy explicit

* Triggering Ansible bot due to timeout

* Updated with changes from comments

migumun committed:
Fortinet's FortiOS wanopt settings (#52840)

* Fortinet's FortiOS wanopt settings

* Updated with changes from comments

migumun committed:
Fortinet's FortiOS wanopt profile (#52839)

* Fortinet's FortiOS wanopt profile

* Updated with changes from comments

migumun committed:
Fortinet's FortiOS waf profile (#52834)

* Fortinet's FortiOS waf profile

* Updated with changes from comments

NilashishC committed:
New module for BGP configuration management in Arista EOS (#52722)

* New module for BGP in EOS

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Add function to validate input

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Fix line indentation

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Add integration tests

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Fix CI

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Fix sanity test failure

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Remove unused code

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

NilashishC committed:
New module for BGP configuration management in iosxr (#53121)

* Add iosxr_bgp module

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Remove unused code

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

* Remove netconf code

Signed-off-by: NilashishC <nilashishchakraborty8@gmail.com>

krisvasudevan committed:
Added support for infoblox member object (#53174)

* changes for member module

dagwieers committed:
Fix typo

chrisvanheuveln committed:
nxos_vxlan_vtep: sanity.yaml test needs cast for httpapi tcam check (#53531)

This check was working for network_cli (returns value as int) but
failing for httpapi (returns value as str). Casting to int fixes the problem.

chrisvanheuveln commented:
I'm looking into the unit test failure.

chrisvanheuveln committed:
nxos_static_route: UT setup changes

`reconcile_candidate()` looks up existing `routes` on the device. These route configs are no longer indented.

chrisvanheuveln committed:
nxos_static_route: UT setup changes

`reconcile_candidate()` looks up existing `routes` on the device. These route configs are no longer indented.

sivel committed:
Don't raise AnsibleConnectionFailure if the ssh process has already died. (#53534)

* Don't raise AnsibleConnectionFailure if the ssh_process has already died. Fixes #53487

* Better support for file not found messages

* Add changelog fragment

bcoca committed:
Overridable safety (#53458)

* create overridable sanitation function
* now used in aws, gce and azure plugins
* added new option to these plugins to work in conjunction with general toggle to make it easier
  to emulate inventory script behavior.

mikewiebe commented on test/units/modules/network/nxos/fixtures/nxos_static_route.cfg:
```diff
@@ -1,3 +1,3 @@
 ip route 10.10.30.0/24 1.2.4.8
 vrf context test
-  ip route 10.8.0.0/14 15.16.17.18
+ip route 10.8.0.0/14 15.16.17.18
```
How do the new changes differentiate between routes under the default and non-default vrf?

chrisvanheuveln replied:
I found a flaw in my implementation and need to rework this a bit. Let's discuss offline. Thanks.

gaudenz committed:
Integration test cleanup for cloudscale modules (#53629)

* Integration test cleanup for cloudscale modules

This refactors the cleanup procedure for the integration tests of the
cloudscale_server and cloudscale_volume modules to use an "always"
section for cleanup. The cleanup code also deletes all resources which
contain the test run prefix. This ensures that all resources are cleaned
up regardless of the actual test result which is a prerequisite for
running these tests in CI.

* Move cloudscale_floating_ip tests from legacy to integration

This also adds code to make sure that floating IPs are deleted even if a
test run fails. This is unfortunately not possible for floatint IPv6
networks.

sivel committed:
Fix checksum file parsing in get_url (#53685)

* Fix checksum file parsing. Fixes #48790

* guard invalid int conversion

Co-Authored-By: sivel <matt@sivel.net>

* Remove extra newline

mkrizek committed:
yum: fix disable_excludes in repoquery fallback (#53552)

Fixes #53134

mkrizek committed:
Add changelog for #53552 (#53695)

migumun committed:
Fortinet's FortiOS log webtrends filter (#52272)

* Fortinet's FortiOS log webtrends filter

* Avoid null choices

* Updated according to comments in PR

* Removed empty choices

migumun committed:
Fortinet's FortiOS system virtual wan link (#52830)

* Fortinet's FortiOS system virtual wan link

* Updated according to comments

migumun committed:
Fortinet's FortiOS system global (#52826)

* Fortinet's FortiOS system global

* Updated according to comments

mattclay committed:
Avoid auto-interpreter for ansible-pull localhost. (#53699)

bcoca committed:
clear all loader caches with new dir for plugin (#53413)

* clear all loader caches with new dir for plugin

  fixes #17078

* added toggle to revert behaviour

mattclay committed:
Disable old_style_cache_plugins integration test.

gaudenz committed:
Cloudscale server password (#53701)

* cloudscale_server: remove required parameter check

The valid parameter combinations are already checked by the API. There
is no need to check a subset of this in the Ansible module as well. This
makes it more likely that future changes in the API won't require
changes in the Ansible module.

* cloudscale_server: add password parameter

Recent revisions of the cloudscale.ch API allow setting a password for a
server. Add this option to the cloudscale_server module.

mattclay committed:
Disable failing azure_rm_keyvault test.

bcoca committed:
fixes for stripping (#52930)

  function changed to do in place replacement, should be less expensive even with copy as it avoids 'sub copies', can compose with module_args_copy to create replacement for old behavior

  attempt to fix #52910

* handle lists and subdicts correctly
* added  missing exception case, which was not noticed since 'cleaning' was not working
* added comments to clarify exceptions

dpassante committed:
cloudstack: streamline modules doc (part 3) (#53412)

* cloudstack: streamline modules doc (part 3)

* Parameter types added
* Copyright format fixes
* Doc style fixes
* Examples format fixes
* Remove cs_loadbalancer_rule from ignore.txt

* cs_loadbalancer_rule: add "version_added" for params added into the documentation

* cs_loadbalancer_rule: ignore E309 for now

* cs_loadbalancer_rule: remove E309 from ignore.txt

jborean93 committed:
win_reboot - Fix rc validation when using psrp and add extra docs (#53711)

* win_reboot - Fix rc validation when using psrp and add extra docs

* Revert boot time command and fix docs

jborean93 committed:
win_reboot: fix stray entry in module docs

jborean93 committed:
win_psexec: make the tests more stable (#53716)

jwitko committed:
Fix min_size reference where type is None (#53669)

jborean93 committed:
win_reboot - fix broken tests after recent commit (#53722)

jborean93 committed:
Use shared remote_tmp_dir to simplfy tests (#53721)

chrisvanheuveln committed:
nxos_linkagg: `group` type mismatch causes idempotency failure (#53653)

* nxos_linkagg: `group` type mismatch causes idempotency failure

* `group` values need to be cast; e.g.

```
 want = {'group': '20'}
 have = {'group': 20}
```

* Found with N7K `sanity` test

* nxos_linkagg: change group param type to str

sumkincpp committed:
vmware_vm_facts: custom attributes support (#45773)

* Add toggle to attribute facts gathering

Signed-off-by: Fedor V <f.vompe@comptek.ru>
Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

migumun committed:
Fortinet's FortiOS voip profile (#52833)

* Fortinet's FortiOS voip profile

* Remove 'flatten' method if not used

felixfontein committed:
acme_challenge_cert_helper: add support for IP identifiers (#53661)

* Add IP address identifier support to acme_challenge_cert_helper.

* Add changelog.

* type -> identifier_type.

felixfontein committed:
ACME: add support for IP identifiers (#53660)

* Adding support for IP identifiers according to https://tools.ietf.org/html/draft-ietf-acme-ip-05.

* Add changelog.

* Make sure that the authorizations return value is unchanged for CSRs with DNS-only SANs.

* Remove unneeded import.

* type -> identifier_type

* Python 2.6 compatibility.

* Fix unit tests.

* Add IP address normalization.

* Extend tests.

* Move data into fixtures.

* Adjust BOTMETA.

justjais committed:
PR to include support for Skydive Node and Edge modules with Ansible (#53112)

* skydive node and edge module

Signed-off-by: Sumit Jaiswal <sjaiswal@redhat.com>

mmazur committed:
kubevirt_pvc: add 'wait' support (#53692)

Akasurde committed:
iptables: Add support for gateway parameter (#53465)

When user specifies the JUMP value to 'tee', gateway is required.
This fix adds new parameter 'gateway' to support this functionality.

Fixes: #53170

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Andersson007 committed:
postgresql_db - Handle pg_dump return code (#52985)

Handle return code return by pg_dump command

Fixes: #40424

Akasurde committed:
iptables: implement log_level parameter (#52880)

Fixes: #25100

Based upon https://github.com/ansible/ansible/pull/25118

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

rambleraptor committed:
give GCP team access to the GCP document fragment (#53659)

carchi8py committed:
Modify na_ontap_net_port NetAppModule module format, allow for multiple ports, unit tests (#52286)

* Revert "changes to clusteR"

This reverts commit 33ee1b71e4bc8435fb315762a871f8c4cb6c5f80.

* add update to netport

* Revert "Revert "changes to clusteR""

This reverts commit 8e56b999e6cf6a65de339e516f7134a6b6b39cba.

* fix docs

* fix option

Akasurde committed:
java_cert: Remove redundant return (#52879)

* Remove redundant return statements for module.exit_json
* Doc changes
* Typo fixes
* Refactor for check_mode

Based upon work of pilou-

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

mattclay committed:
Overhaul ansible-test code coverage and injector. (#53510)

alucas committed:
Add headers parameter to the maven_artifact module (#53116)

* Fix header parameter type

sivel committed:
cliconf and httpapi are documentable plugins now

erav committed:
ovirt: default save true in setup host networks (#49983)

* ovirt: default save true in setup host networks

Default value for save updated network configuration
during setup host networks is set to true.

* ovirt: default save true in setup host networks

Default value for save updated network configuration
during setup host networks is set to true.

* ovirt: default save true in setup host networks

Default value for save updated network configuration
during setup host networks is set to true.

* Update ovirt_host_network.py

* ovirt: default save true in setup host networks

Default value for save updated network configuration
during setup host networks is set to true.

* ovirt: default save true in setup host networks

Default value for save updated network configuration
during setup host networks is set to true.

rambleraptor committed:
GCP IAM Role (#53490)

* GCP IAM Role

* module util file

* test fix

* unit tests

* test fixes

* doc fragment fixes

* test fixes

* test fix

dpassante committed:
cs_iso: fix missing param "is_public" (#53740)

* cs_iso: fix missing param "is_public"

* add changelog fragment

chrisvanheuveln committed:
nxos_static_route: strip leading spaces from ip route configs

chrisvanheuveln committed:
nxos_static_route: strip leading spaces from ip route configs

mmazur committed:
More kubevirt_* modules in k8s mod defaults group (#53450)

albinpopote committed:
na_ontap_dns: Add new parameter skip_validation (#50458)

* Add new parameter skip_validation

* Change file indentation

* Indentation fix

* Change module documentation

* Contributor change request : fixed

* FIX : Validator change request

* Comment: Add more description details about skip_validation option

albinpopote committed:
New module: na_ontap_net_subnet (#50491)

* New module: na_ontap_net_subnet

* Sanity tests Fix

* Fix: correct examples

* Fix: examples syntax

* Correction: Sanity 1

* Code review

* Code review + Bugfix

* Add Unit Tests

* Fix Sanity Indentation tests

* FIX: Validator change request

* FIX : Change unit test json fail message

* FIX: Typo ip_ranges in example section

* FIX: Remove NetApp copyright

carchi8py committed:
Merging of broadcast domain and Broadcast domain port (#51978)

* Revert "changes to clusteR"

This reverts commit 33ee1b71e4bc8435fb315762a871f8c4cb6c5f80.

* Add domain ports in to broadcast domain

* ansibot fixes

* Revert "Revert "changes to clusteR""

This reverts commit 2713c75f31cbf81ef1785d6ab9ea5d0d4db9af60.

* Fix author

gabriel376 committed:
Add non_unique parameter for group module (#53085)

bcoca committed:
Undeprecate force handlers (#53705)

* removed deprecation notice, made comment instead

* remove noisy deprecation

* space

chrisvanheuveln committed:
Merge branch 'devel-nxos_static_route' of https://github.com/chrisvanheuveln/ansible into devel-nxos_static_route
